### PR TITLE
[Power8] Add chunk*_power8.

### PR DESCRIPTION
**Pull request by mtl1979**
VSX optimized chunk copy and memset functions.

### CODE REVIEW

File: arch/power/chunkset_power8.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+#define CHUNKSIZE        chunksize_power8
+#define CHUNKCOPY        chunkcopy_power8
+#define CHUNKCOPY_SAFE   chunkcopy_safe_power8
+#define CHUNKUNROLL      chunkunroll_power8
+#define CHUNKMEMSET      chunkmemset_power8
+#define CHUNKMEMSET_SAFE chunkmemset_safe_power8
+
+static inline void loadchunk(uint8_t const *s, chunk_t *chunk) {
+    *chunk = vec_vsx_ld(0, s);
```
Comment (mscastanho) [inline arch/power/chunkset_power8.c]: It's better to use `vec_xl` as it is equivalent but more portable.

File: arch/power/chunkset_power8.c [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+static inline void storechunk(uint8_t *out, chunk_t *chunk) {
+    vec_vsx_st(*chunk, 0, out);
```
Comment (mscastanho) [inline arch/power/chunkset_power8.c]: It's better to use `vec_xst` as it is equivalent but more portable.

File: arch/power/chunkset_power8.c (hunk 1 above)
Comment (mtl1979) [reply arch/power/chunkset_power8.c]: Changed...

File: arch/power/chunkset_power8.c (hunk 2 above)
Comment (mtl1979) [reply arch/power/chunkset_power8.c]: Changed.

File: arch/power/chunkset_power8.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
+/* chunkset_power8.c -- VSX inline functions to copy small data chunks.
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#ifdef POWER8_VSX_CHUNKSET
+#include <altivec.h>
+#include "../../zbuild.h"
+#include "../../zutil.h"
```
Comment (mscastanho) [inline arch/power/chunkset_power8.c]: No need to make includes relative because the proper `-I` is passed to the compiler. Just remove the `../../`

**Review by mscastanho — COMMENTED** [review]
@mtl1979 LGTM. I just suggested a small change.

But I could not measure a significant performance increase with this change. On a POWER8:

Develop (c0cca0854206e25b)
```
Tool: minigzip   Levels: 0-9       
 Runs: 15         Trim worst: 5         

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.671/0.698/0.727/0.020        0.639/0.673/0.707/0.019      536,124,384
 1     54.240%      2.397/2.417/2.436/0.014        1.097/1.143/1.159/0.019      154,205,532
 2     43.987%      2.118/2.137/2.149/0.011        0.567/0.597/0.616/0.016       62,758,792
 3     42.526%      2.053/2.075/2.087/0.011        0.436/0.451/0.466/0.013       47,290,047
 4     41.837%      1.882/1.910/1.925/0.012        0.349/0.379/0.394/0.013       39,939,949
 5     41.384%      1.895/1.919/1.933/0.010        0.358/0.373/0.389/0.011       39,507,653
 6     41.246%      1.784/1.799/1.816/0.010        0.293/0.317/0.330/0.010       32,885,139
 7     41.044%      1.795/1.805/1.816/0.006        0.254/0.258/0.264/0.004       26,265,857
 8     41.053%      1.656/1.676/1.687/0.009        0.177/0.195/0.207/0.011       19,811,261
 9     41.043%      1.796/1.822/1.829/0.011        0.177/0.193/0.207/0.010       19,806,112

 avg1  48.837%                        1.826                          0.458
 avg2  54.263%                        2.029                          0.509
 tot                                182.591                         45.803      978,594,726
```

This PR:
```
 Tool: minigzip   Levels: 0-9
 Runs: 15         Trim worst: 5

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.669/0.697/0.717/0.017        0.611/0.663/0.699/0.028      536,124,384
 1     54.240%      2.386/2.415/2.437/0.016        1.110/1.135/1.157/0.017      154,205,532
 2     43.987%      2.104/2.132/2.145/0.015        0.544/0.579/0.599/0.018       62,758,792
 3     42.526%      2.036/2.067/2.084/0.015        0.418/0.446/0.462/0.014       47,290,047
 4     41.837%      1.890/1.919/1.931/0.012        0.364/0.382/0.395/0.009       39,939,949
 5     41.384%      1.899/1.910/1.921/0.008        0.353/0.382/0.391/0.012       39,507,653
 6     41.246%      1.771/1.787/1.803/0.012        0.293/0.305/0.315/0.008       32,885,139
 7     41.044%      1.795/1.813/1.824/0.008        0.239/0.249/0.259/0.008       26,265,857
 8     41.053%      1.636/1.663/1.676/0.015        0.172/0.185/0.192/0.007       19,811,261
 9     41.043%      1.808/1.828/1.837/0.011        0.181/0.193/0.202/0.008       19,806,112

 avg1  48.837%                        1.823                          0.452
 avg2  54.263%                        2.026                          0.502
 tot                                182.311                         45.191      978,594,726
```

So it is very slightly better, but it's hard to distinguish from noise.

File: arch/power/chunkset_power8.c (hunk 3 above)
Comment (mtl1979) [reply arch/power/chunkset_power8.c]: Changed.

File: arch/power/chunkset_power8.c [hunk 4]
```diff
@@ -0,0 +1,58 @@
+/* chunkset_power8.c -- VSX inline functions to copy small data chunks.
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#ifdef POWER8_VSX_CHUNKSET
+#include <altivec.h>
+#include "zbuild.h"
+#include "/zutil.h"
```
Comment (mscastanho) [inline arch/power/chunkset_power8.c]: @mtl1979 you forgot the leading `/`

Comment (mtl1979) [reply arch/power/chunkset_power8.c]: ;)

**Comment by mtl1979** [comment]
@mscastanho Sometimes the performance gain from using AltiVec isn't large on all processors... I noticed later processors can implement vec_xl/vec_xst better (single instruction instead of two).

**Review by mscastanho — APPROVED** [review]
@mtl1979 You're right. I ran a benchmark on a POWER9, but I couldn't use `deflatebench` this time, not sure why. I kept getting `0.000` on all measurements. So I used the benchmark I used for #603 and #647. Looks like the PR improves performance in some cases. So LGTM.

```
======== File 'pngpixels' ======== 

 Wrapper 'gzip'

default
          |compression throughput (MB) |decompression throughput (MB)|
  | level | default | optim |     gain | default | optim |     gain |
  |     1 |  178.4  | 178.7 |   +0.17% |  258.0  | 276.2 |   +7.05% |
  |     2 |  121.4  | 121.0 |   -0.33% |  270.9  | 286.0 |   +5.57% |
  |     3 |  89.6   |  88.1 |   -1.67% |  280.4  | 298.3 |   +6.38% |
  |     4 |  70.3   |  73.6 |   +4.69% |  298.3  | 320.2 |   +7.34% |
  |     5 |  61.4   |  62.3 |   +1.47% |  307.0  | 330.4 |   +7.62% |
  |     6 |  36.2   |  36.1 |   -0.28% |  326.7  | 353.3 |   +8.14% |
  |     7 |  20.9   |  20.9 |   +0.00% |  323.9  | 346.9 |   +7.10% |
  |     8 |   6.7   |   6.7 |   +0.00% |  329.4  | 351.9 |   +6.83% |
  |     9 |   3.1   |   3.1 |   +0.00% |  331.2  | 353.3 |   +6.67% |

======== File 'jpeg' ======== 

 Wrapper 'gzip'

default
          |compression throughput (MB) |decompression throughput (MB)|
  | level | default | optim |     gain | default | optim |     gain |
  |     1 |  47.1   |  47.6 |   +1.06% |  156.1  | 155.9 |   -0.13% |
  |     2 |  33.1   |  33.4 |   +0.91% |  540.8  | 540.1 |   -0.13% |
  |     3 |  29.5   |  29.4 |   -0.34% |  540.7  | 540.1 |   -0.11% |
  |     4 |  29.6   |  29.6 |   +0.00% |  540.7  | 540.0 |   -0.13% |
  |     5 |  25.8   |  25.8 |   +0.00% |  540.6  | 539.7 |   -0.17% |
  |     6 |  25.6   |  25.8 |   +0.78% |  540.3  | 539.9 |   -0.07% |
  |     7 |  32.0   |  31.9 |   -0.31% |  540.9  | 540.2 |   -0.13% |
  |     8 |  32.0   |  31.6 |   -1.25% |  540.6  | 540.2 |   -0.07% |
  |     9 |  32.0   |  32.0 |   +0.00% |  540.5  | 540.0 |   -0.09% |

======== File 'executable' ======== 

 Wrapper 'gzip'

default
          |compression throughput (MB) |decompression throughput (MB)|
  | level | default | optim |     gain | default | optim |     gain |
  |     1 |  111.3  | 111.3 |   +0.00% |  198.3  | 208.2 |   +4.99% |
  |     2 |  68.2   |  67.6 |   -0.88% |  185.6  | 196.1 |   +5.66% |
  |     3 |  57.2   |  56.9 |   -0.52% |  190.3  | 201.9 |   +6.10% |
  |     4 |  50.0   |  50.0 |   +0.00% |  192.9  | 205.4 |   +6.48% |
  |     5 |  45.6   |  45.7 |   +0.22% |  194.8  | 208.0 |   +6.78% |
  |     6 |  38.8   |  38.8 |   +0.00% |  195.8  | 209.3 |   +6.89% |
  |     7 |  23.1   |  23.1 |   +0.00% |  196.5  | 210.4 |   +7.07% |
  |     8 |  15.1   |  15.1 |   +0.00% |  196.9  | 210.9 |   +7.11% |
  |     9 |   9.7   |   9.7 |   +0.00% |  197.0  | 211.2 |   +7.21% |

======== File 'html' ======== 

 Wrapper 'gzip'

default
          |compression throughput (MB) |decompression throughput (MB)|
  | level | default | optim |     gain | default | optim |     gain |
  |     1 |  111.7  | 112.3 |   +0.54% |  189.3  | 199.8 |   +5.55% |
  |     2 |  69.8   |  69.2 |   -0.86% |  202.4  | 217.7 |   +7.56% |
  |     3 |  58.0   |  58.0 |   +0.00% |  211.6  | 230.1 |   +8.74% |
  |     4 |  47.4   |  47.3 |   -0.21% |  214.7  | 236.9 |  +10.34% |
  |     5 |  43.7   |  43.6 |   -0.23% |  213.7  | 237.1 |  +10.95% |
  |     6 |  36.1   |  36.1 |   +0.00% |  214.0  | 238.7 |  +11.54% |
  |     7 |  24.0   |  24.0 |   +0.00% |  208.0  | 233.5 |  +12.26% |
  |     8 |  16.7   |  16.7 |   +0.00% |  208.1  | 233.4 |  +12.16% |
  |     9 |  16.4   |  16.4 |   +0.00% |  208.0  | 233.5 |  +12.26% |
```

**Comment by mscastanho** [comment]
I also tested the code on ppc and ppc64 (BE). Code builds fine and all tests pass.

**Comment by mtl1979** [comment]
@mscastanho I'm not sure if minimum is Power7 or Power8 for the current implementation, but we don't have feature test for Power7, so using Power8 as minimum is acceptable, especially if code works best with Power9 or later...

**Comment by mscastanho** [comment]
@mtl1979 I think it's fine to leave P8 as the baseline. zlib-ng is targeting "next generation systems", and P7 is already 10+ years old and it's harder to get hold of one for testing. This same optimization will be used by P8 and later, so they all get the benefit from this.

I even did a quick benchmark compiling `chunkset_power8` with `-mcpu=power9` to see if we would get better performance. It is very slightly better (< 1%), so I don't think it's worth adding a P9 feature check just for it.

**Comment by mtl1979** [comment]
@mscastanho Yeah... Even though I like BeBoxes, it's already 25 years old technology... lol... but good for learning about PowerPC architecture and its differences compared to 80x86 and ARM... I got amused first time when I booted Ubuntu on PowerPC and it was only using 200 MB of 2 GB installed main memory... Can't really do same with Intel or AMD...

zlib-ng isn't always all about latest hardware as it still supports 32-bit processors ;)

**Review by nmoinvaz — APPROVED** [review]

**Comment by mtl1979** [comment]
@Dead2 Should we merge this for 2.0.6 or... Might be already too much rushing to get for 2.0.5 now...

**Comment by Dead2** [comment]
No, the stable 2.0.x branch is closed for features. This will go to develop and be released in 2.1.0.

**Comment by mtl1979** [comment]
@Dead2 I'm ok with that... There will be more PowerPC improvements anyways...

**Pull request merged**